macro/macro.7
.TH YED-MACRO 7 "YED Plugin Manuals" "" "YED Plugin Manuals"
.SH NAME
macro \- Record keystrokes and play them back.
.SH CONFIGURATION
.SS macro-stop-key [key-string]
The key designated to mean "stop recording the macro".
Default value is ctrl-e.
.SS macro-instant-playback [ yes / no ]
When this variable is set to "yes", keystrokes will be replayed instantly -- i.e. you won't see them happen.
Default value is "no".
Otherwise keystrokes will be replayed once per editor pump.
.SH COMMANDS
.SS macro-record
Start recording a macro.
.SS macro-play [ N ]
If N is provided, play the macro N times.
Otherwise play the macro once for each line selected (once if no selection).
.SH BUFFERS
None
.SH NOTES
None
.SH VERSION
0.0.1
.SH KEYWORDS
key, record, macro, play, playback, repeat


CHANGELOG.md
# Changelog

## 1306 - 2021-12-13
### Fixed
    - fixed a bug that would cause command line contents to overflow into a new terminal line.

## 1305 - 2021-12-13
### Fixed
    - fixed a bug that would cause some cursor movements to crash on binary buffer contents.

## 1304 - 2021-12-10
### Added
    - new command `frame` activates frame based on their number.

## 1303 - 2021-12-8
### Changed
    - the `plugins` directory is now a `subtree`. This makes installing simpler and allows you to just grab a `tar` of a release without needing `git`.

## 1302 - 2021-12-7
### Changed
    - removed use of `codesign` in the install script for Darwin targets. This isn't needed any more and will help us package `yed`.

## 1301 - 2021-12-7
### Fixed
    - Fixed logic in yed_cell_is_in_frame()

## 1300 - 2021-12-6
### Added
- Started keeping a changelog.
- Started tagging versions in git.
- A new syntax highlighting engine in `syntax.h`.
- Status line additions:
    - `%{...}` syntax for expanding a variable that contains an attribute string
    - Padding and justification on `%` expansions: `%` followed by an integer, `x`, will be padded to `x` characters. If `x` is preceded by a `-`, the text will be left justified. A `=` indicates center justification. The default is right justification. This has no effect on expansions that result in attributes.
    - `%T`: 24-hour clock
- Events:
    - `EVENT_STATUS_LINE_PRE_UPDATE` will trigger before the status line is about to be updated.
    - `EVENT_PLUGIN_MESSAGE` is a new general purpose event that plugins can send and other plugins can handle.
- Mouse support added to the core for left, center, and right buttons as well as up and down scroll wheel. Distinguishes between press, release, and drag.
- Useful functions for getting buffer contents as byte strings:
    - `char *yed_get_selection_text(yed_buffer *buffer);`
    - `char *yed_get_line_text(yed_buffer *buffer, int row);`
    - `char *yed_get_buffer_text(yed_buffer *buffer);`
- Style components:
    - `good`: general indication in the positive (e.g. a notification that a build succeeded)
    - `bad`: general indication in the negative (e.g. a notification that a build failed)
    - `popup`: the foreground and background to be used for popup menus
    - `popup-alt`: the foreground and background to be used for other popup menu elements (e.g. the selected item)
    - `code-variable`: for variables in code
    - `code-field`: for fields of a structure or class
    - `code-escape`: for escape sequences in string literals
### Changed
- `highlight.h` is now deprecated and will be removed in version 1400.
- Status line: `%T`, which used to expand to the filetype, now expands to a 24-hour clock. `%F` now expands to the filetype.
### Removed
- Removed the default command `if`.
### Fixed
- A buffer's undo history is now reset when `buffer-reload` gets run (or whenever the contents of the buffer are loaded from a file).
- Plugins are now install-agnostic -- they will use whatever `libyed.so` is being used by the binary. This fixes a crash that would occur when installations would mix on the same system.


plugins/macro.c
#include <yed/plugin.h>

typedef struct {
    int  n;
    int  keys[16];
} key_seq_t;

yed_plugin        *Self;
yed_event_handler  rec_key_handler;
yed_event_handler  play_key_handler;
yed_event_handler  pump_rec_handler;
yed_event_handler  pump_play_handler;
int                is_recording;
int                has_recorded;
int                stop_key;
char              *stop_key_str;
array_t            keys;
int                playback_idx;
int                blink_count;
yed_direct_draw_t *blink_dd;
u64                blink_start_time;
int                play_count;
int                has_played_count;
int                key_pressed_is_playback;

void macro_record(int n_args, char **args);
void macro_play(int n_args, char **args);

void macro_rec_key_handler(yed_event *event);
void macro_play_key_handler(yed_event *event);
void macro_pump_rec_handler(yed_event *event);
void macro_pump_play_handler(yed_event *event);

void macro_unload(yed_plugin *self);

int yed_plugin_boot(yed_plugin *self) {
    YED_PLUG_VERSION_CHECK();

    Self = self;

    rec_key_handler.kind   = EVENT_KEY_PRESSED;
    rec_key_handler.fn     = macro_rec_key_handler;
    play_key_handler.kind  = EVENT_KEY_PRESSED;
    play_key_handler.fn    = macro_play_key_handler;
    pump_rec_handler.kind  = EVENT_PRE_PUMP;
    pump_rec_handler.fn    = macro_pump_rec_handler;
    pump_play_handler.kind = EVENT_PRE_PUMP;
    pump_play_handler.fn   = macro_pump_play_handler;

    if (yed_get_var("macro-stop-key") == NULL) {
        yed_set_var("macro-stop-key", "ctrl-e");
    }
    if (yed_get_var("macro-instant-playback") == NULL) {
        yed_set_var("macro-instant-playback", "no");
    }

    keys = array_make(key_seq_t);

    yed_plugin_set_command(self, "macro-record", macro_record);
    yed_plugin_set_command(self, "macro-play",   macro_play);

    return 0;
}

void macro_unload(yed_plugin *self) {
    array_free(keys);
}

void macro_rec_key_handler(yed_event *event) {
    key_seq_t seq;

    if (event->key == stop_key) {
        yed_delete_event_handler(rec_key_handler);
        yed_delete_event_handler(pump_rec_handler);
        if (blink_dd != NULL) {
            yed_kill_direct_draw(blink_dd);
            blink_dd = NULL;
        }
        has_recorded  = 1;
        is_recording  = 0;
        event->cancel = 1;
    } else {
        if (!yed_get_real_keys(event->key, &seq.n, seq.keys)) {
            seq.n       = 1;
            seq.keys[0] = event->key;
        }
        array_push(keys, seq);
    }
}

void macro_play_key_handler(yed_event *event) {
    if (key_pressed_is_playback) { return; }

    if (event->key == CTRL_C) {
        if (blink_dd != NULL) {
            yed_kill_direct_draw(blink_dd);
            blink_dd = NULL;
        }

        yed_delete_event_handler(pump_play_handler);
        yed_delete_event_handler(play_key_handler);
LOG_CMD_ENTER("macro-play");
        yed_cprint("finished playing back macro");
LOG_EXIT();
    }

    event->cancel = 1;
}

void macro_pump_rec_handler(yed_event *event) {
    u64       now;
    yed_attrs attrs;
    char      buff[64];

    if (blink_dd != NULL) {
        yed_kill_direct_draw(blink_dd);
        blink_dd = NULL;
    }

    now = measure_time_now_ms();

    if (now - blink_start_time > 1000) {
        blink_start_time = now;
        blink_count += 1;
    }

    if (blink_count % 2 == 0) {
        snprintf(buff, sizeof(buff), " macro recording    %s to stop ", stop_key_str);
        attrs        = yed_active_style_get_attention();
        attrs.flags |= ATTR_INVERSE;
        blink_dd     = yed_direct_draw(ys->term_rows - 2, 1, attrs, buff);
    }
}

void macro_play_key_seq(key_seq_t *seq) {
    int virt;

    key_pressed_is_playback = 1;
    if (seq->n == 1
    ||  (virt = yed_get_key_sequence(seq->n, seq->keys)) == KEY_NULL) {
        yed_feed_keys(1, seq->keys);
    } else {
        yed_feed_keys(1, &virt);
    }
    key_pressed_is_playback = 0;
}

void macro_pump_play_handler(yed_event *event) {
    u64       now;
    yed_attrs attrs;

    if (blink_dd != NULL) {
        yed_kill_direct_draw(blink_dd);
        blink_dd = NULL;
    }

    now = measure_time_now_ms();

    if (blink_count % 2 == 0) {
        attrs        = yed_active_style_get_associate();
        attrs.flags |= ATTR_INVERSE;
        blink_dd     = yed_direct_draw(ys->term_rows - 2, 1, attrs,
                                       " playing back macro    ctrl-c to cancel ");
    }

    if (now - blink_start_time > 1000) {
        blink_start_time = now;
        blink_count += 1;
    }

    if (playback_idx == array_len(keys)) {
        has_played_count += 1;
        if (has_played_count == play_count) {
            if (blink_dd != NULL) {
                yed_kill_direct_draw(blink_dd);
                blink_dd = NULL;
            }

            yed_delete_event_handler(pump_play_handler);
            yed_delete_event_handler(play_key_handler);
LOG_CMD_ENTER("macro-play");
            yed_cprint("finished playing back macro");
LOG_EXIT();
        } else {
            playback_idx = 0;
        }
    } else {
        macro_play_key_seq(array_item(keys, playback_idx));
        playback_idx += 1;
    }
}

void macro_record(int n_args, char **args) {
    char *key_str;
    char  key_c;
    int   key_i;

    if (is_recording) {
        yed_cerr("a macro is currently being recorded");
        return;
    }

    if (n_args > 1) {
        yed_cerr("expected 0 or 1 arguments, but got %d", n_args);
        return;
    }

    if (n_args == 1) {
        key_str = args[0];
    } else {
        key_str = yed_get_var("macro-stop-key");
    }

    if (key_str == NULL) {
        yed_set_var("macro-stop-key", "ctrl-e");
    }

    key_c = key_i = -1;

    if (strlen(key_str) == 1) {
        sscanf(key_str, "%c", &key_c);
        key_i = key_c;
    } else if (strcmp(key_str, "tab") == 0) {
        key_i = TAB;
    } else if (strcmp(key_str, "enter") == 0) {
        key_i = ENTER;
    } else if (strcmp(key_str, "esc") == 0) {
        key_i = ESC;
    } else if (strcmp(key_str, "spc") == 0) {
        key_i = ' ';
    } else if (strcmp(key_str, "bsp") == 0) {
        key_i = BACKSPACE;
    } else if (sscanf(key_str, "ctrl-%c", &key_c)) {
        if (key_c != -1) {
            if (key_c == '/') {
                key_i = CTRL_FS;
            } else {
                key_i = CTRL_KEY(key_c);
            }
        }
    }

    if (key_i == -1) {
        yed_cerr("invalid macro-stop-key '%s'", key_str);
        return;
    }

    stop_key = key_i;
    if (has_recorded) { free(stop_key_str); }
    stop_key_str = strdup(key_str);
    is_recording = 1;
    blink_count  = 0;
    blink_start_time = measure_time_now_ms();
    array_clear(keys);
    yed_plugin_add_event_handler(Self, rec_key_handler);
    yed_plugin_add_event_handler(Self, pump_rec_handler);
}

void macro_play(int n_args, char **args) {
    int        i;
    key_seq_t *seq_it;
    yed_range *sel;

    if (is_recording) {
        yed_cerr("a macro is currently being recorded");
        return;
    }

    if (n_args > 1) {
        yed_cerr("expected 0 or 1 arguments, but got %d", n_args);
        return;
    }

    if (n_args == 0
    ||  sscanf(args[0], "%d", &play_count) != 1
    ||  play_count < 0) {

        if (ys->active_frame
        &&  ys->active_frame->buffer
        &&  ys->active_frame->buffer->has_selection) {
            sel = &(ys->active_frame->buffer->selection);
            play_count = abs(sel->anchor_row - sel->cursor_row) + 1;
            yed_set_cursor_within_frame(ys->active_frame,
                                         sel->anchor_row, sel->anchor_col);
            ys->active_frame->buffer->has_selection = 0;
        } else {
            play_count = 1;
        }

    }

    blink_count = 0;

    if (yed_var_is_truthy("macro-instant-playback")) {
        for (i = 0; i < play_count; i += 1) {
            array_traverse(keys, seq_it) {
                macro_play_key_seq(seq_it);
            }
        }
    } else {
        playback_idx     = 0;
        has_played_count = 0;
        yed_plugin_add_event_handler(Self, pump_play_handler);
        yed_plugin_add_event_handler(Self, play_key_handler);
    }
}
